R/tools.R

```diff
@@ -19,6 +19,84 @@ test_r <- function(...) {
   )
 }
 
+# From testthat https://github.com/r-lib/testthat/blob/v3.2.3/R/auto-test.R#L130
+starts_with <- function(string, prefix) {
+  substr(string, 1, nchar(prefix)) == prefix
+}
+
+check_if_includes_r_files <- function(path) {
+  r_files <- fs::dir_ls(path, recurse = TRUE, glob = "*.R$")
+
+  if (length(r_files) > 0) {
+    return(normalizePath(path))
+  }
+
+  NULL
+}
+
+# nolint start: line_length_linter
+#' Watch and automatically run R tests
+#'
+#' Watches R files in the `app` directory and `tests/testthat` directory for changes.
+#' When code files in `app` change, all tests are rerun. When test files change,
+#' only the changed test file is rerun.
+#'
+#' @param reporter `{testthat}` reporter to use.
+#'   If NULL, will use `testthat::default_reporter()` for tests when running all tests
+#'   and `testthat::default_compact_reporter()` for single file tests.
+#'   See [`{testthat}` reporters](https://testthat.r-lib.org/articles/reporters.htmlhttps://testthat.r-lib.org/reference/Reporter.html) for more details.
+#' @param hash Logical. Whether to use file hashing to detect changes. Default is TRUE.
+#'   If FALSE, file modification times are used instead.
+#'
+#' @return None. This function is called for side effects.
+#'
+#' @examples
+#' if (interactive()) {
+#'   # Watch files and automatically run tests when changes are detected
+#'   auto_test_r()
+#' }
+#' @export
+# nolint end
+auto_test_r <- function(reporter = NULL, hash = TRUE) {
+  test_path <- normalizePath(fs::path("tests", "testthat"))
+
+  code_path <- c(
+    normalizePath("app"),
+    unlist(fs::dir_map("app", check_if_includes_r_files, recurse = TRUE, type = "directory"))
+  )
+
+  if (is.null(reporter)) {
+    single_file_reporter <- testthat::default_compact_reporter()
+  } else {
+    single_file_reporter <- reporter
+  }
```

Comment:
What's the motivation to supply the default reporter here, instead of as a default parameter?

Reply:
`testthat::test_dir` (used in `rhino::test_r`) and `testthat::test_file` are inconsistent in using `reporter` parameter. In `test_dir`, the deafult is `NULL` and in that case, it will use `thetthat::default_reporter()`. In `test_file`, the default is `default_compact_reporter`, so I had to add this part to provide consistency on our side.

Reply:
I see, that makes sense then. Thank you for explaining it